va/validation-authority.go

```diff
@@ -606,24 +606,46 @@ func newCAASet(CAAs []*dns.CAA) *CAASet {
 func (va *ValidationAuthorityImpl) getCAASet(ctx context.Context, hostname string) (*CAASet, error) {
 	hostname = strings.TrimRight(hostname, ".")
 	labels := strings.Split(hostname, ".")
+
 	// See RFC 6844 "Certification Authority Processing" for pseudocode.
-	// Essentially: check CAA records for the FDQN to be issued, and all parent
-	// domains.
+	// Essentially: check CAA records for the FDQN to be issued, and all
+	// parent domains.
+	//
+	// The lookups are performed in parallel in order to avoid timing out
+	// the RPC call.
+	//
 	// We depend on our resolver to snap CNAME and DNAME records.
+
+	type result struct {
+		records []*dns.CAA
+		err     error
+	}
+	results := make([]result, len(labels))
+
+	var wg sync.WaitGroup
+
 	for i := 0; i < len(labels); i++ {
-		name := strings.Join(labels[i:], ".")
-		// Break if we've reached an ICANN TLD.
-		if tld, err := publicsuffix.ICANNTLD(name); err != nil || tld == name {
-			break
-		}
-		CAAs, err := va.DNSResolver.LookupCAA(ctx, name)
-		if err != nil {
-			return nil, err
+		// Start the concurrent DNS lookup.
+		wg.Add(1)
+		go func(ctx context.Context, name string, r *result) {
```

Comment:
Because `ctx` doesn't change during the loop, it doesn't have to be part of the function arguments. We can just pull it in from the closure's scope.